dvc/remote/ssh/connection.py

```diff
@@ -128,7 +128,17 @@ def walk_files(self, directory):
 
         self._sftp_connect()
 
-        for entry in self._sftp.listdir_attr(directory):
+        try:
+            dir_entries = self._sftp.listdir_attr(directory)
+        except IOError as exc:
+            raise DvcException(
```

Comment:
Other remote implementations like s3 would simply show that directory as empty, which is fine, since we will makedirs all needed directories later when uploading.

Reply:
Not sure if I understand right. Do you suggest to create the directory if it is missing?

Reply:
Or just ignore the exception and return an empty list here?

Reply:
Just ignore the exception and yield nothing, yes :slightly_smiling_face:

Reply:
My 2 cents :) I would log it at least (last time we caught a problem with Windows path), or check a specific type - errno.

Reply:
This could lead to an even more embarrassing error:

```
Error: failed to push data to the cloud - 'host'
```

And traceback:

```
Traceback (most recent call last):
  File "/home/ei-grad/repos/github.com/iterative/dvc/dvc/remote/ssh/__init__.py", line 184, in upload
    from_info["path"], to_info["path"], progress_title=name
  File "/home/ei-grad/repos/github.com/iterative/dvc/dvc/remote/ssh/connection.py", line 181, in upload
    self.makedirs(posixpath.dirname(dest))
  File "/home/ei-grad/repos/github.com/iterative/dvc/dvc/remote/ssh/connection.py", line 121, in makedirs
    self.makedirs(head)
  File "/home/ei-grad/repos/github.com/iterative/dvc/dvc/remote/ssh/connection.py", line 124, in makedirs
    self._sftp.mkdir(path)
  File "/home/ei-grad/.virtualenvs/dvc/lib/python3.7/site-packages/paramiko/sftp_client.py", line 460, in mkdir
    self._request(CMD_MKDIR, path, attr)
  File "/home/ei-grad/.virtualenvs/dvc/lib/python3.7/site-packages/paramiko/sftp_client.py", line 813, in _request
    return self._read_response(num)
  File "/home/ei-grad/.virtualenvs/dvc/lib/python3.7/site-packages/paramiko/sftp_client.py", line 865, in _read_response
    self._convert_status(msg)
  File "/home/ei-grad/.virtualenvs/dvc/lib/python3.7/site-packages/paramiko/sftp_client.py", line 896, in _convert_status
    raise IOError(errno.EACCES, text)
PermissionError: [Errno 13] Permission denied

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/ei-grad/repos/github.com/iterative/dvc/dvc/command/data_sync.py", line 63, in do_run
    recursive=self.args.recursive,
  File "/home/ei-grad/repos/github.com/iterative/dvc/dvc/repo/push.py", line 27, in push
    used, jobs, remote=remote, show_checksums=show_checksums
  File "/home/ei-grad/repos/github.com/iterative/dvc/dvc/data_cloud.py", line 129, in push
    show_checksums=show_checksums,
  File "/home/ei-grad/repos/github.com/iterative/dvc/dvc/remote/local.py", line 704, in push
    download=False,
  File "/home/ei-grad/repos/github.com/iterative/dvc/dvc/remote/local.py", line 694, in _process
    f.result()
  File "/usr/lib64/python3.7/concurrent/futures/_base.py", line 425, in result
    return self.__get_result()
  File "/usr/lib64/python3.7/concurrent/futures/_base.py", line 384, in __get_result
    raise self._exception
  File "/usr/lib64/python3.7/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/ei-grad/repos/github.com/iterative/dvc/dvc/remote/ssh/__init__.py", line 189, in upload
    host=from_info["host"],
KeyError: 'host'
```

Reply:
It is my bad that I didn't explain why the directory doesn't exist :). The actual problem is the lack of clarification about `dvc remote add` on ssh remote that the path has to be absolute path.

Reply:
Also, I've been working on a decorator like this:

```def exception_wrapper(before, after, *argz, logger=None):
    """
    A decorator that catches any exceptions thrown by the decorated function and
    wraps them into a different exception.

    :param before: tuple of exceptions to catch
    :param after: an exception class to raise
    :param argz: arguments to pass to a new exception
    :param logger: (optional) logger to log debug info of the initial exception
    :raises an Exception provided via `after`
    """
    def decorator(func):
        @wraps(func)
        def wrapper(*args, **kwargs):
            try:
                return func(*args, **kwargs)
            except before as e:
                if logger is not None:
                    logger.debug("Exception thrown, {}: {}, in {} with {}"
                                 .format(type(e), str(e), func.__name__, args))
                raise after(*argz, cause=e)
        return wrapper
    return decorator```

and this actual decorator:

```   exception = exception_wrapper(
        (IOError, socket.error, paramiko.SSHException),
        DvcSSHException
```

to wrap all paramiko SSH related calls to avoid catching generic Exception in the caller (like we do now at least in two places).

Just haven't had time to get it done, but idea was to dump the context and write a better message as well. So, probably, don't focus on optimizing the message.

Reply:
I'm pretty sure that this behavior is good for S3, but bad for SSH.

Reply:
@shcheklein idea looks good, but for better UX in the scenario of adding SSH remote there should be a clear error which would show where the things went wrong and maybe suggest the solution like it is done for `git checkout` vs `dvc checkout`.

Reply:
@ei-grad totally, I started working on this when we got an error on Windows and there was no information at all about what'd happened. I pass `cause=e` to the DvcException and it prints `cause.msg` - it seems that should be enough from UX perspective. And verbose mode could print the full stack trace, etc. It looks actually like you solution is not that different in terms of providing a better message. Basically, I was thinking to introduce a more general decorator to solve the same problem + wrap all errors into DvcException.

Reply:
@ei-grad So is the source of this issue relative path in url for ssh remote or is it non-existent directory? If it is the former, then just adding a check that it is an abs path to RemoteSSH.__init__ should be enough for now.

Reply:
Discussed this with @ei-grad in private. I think current patch will suffice for now, as it at least improves experience of finding out the cause for the error(there might be a variety of errors from ENOENT to EPERM, so it makes sense to just catch IOError as a whole). Later @shcheklein will introduce error wrappers and we will consider adding an interactive `dvc remote add` that would try connecting to the remote and verifying that it works while user is configuring it.

Reply:
Sounds good to me! :)